tests/python/unittest/test_higher_order_grad.py

```diff
@@ -210,6 +217,168 @@ def check_second_order_unary(x, op, grad_grad_op, rtol=None, atol=None):
                         x.grad.asnumpy(), rtol=rtol, atol=atol)
 
 
+def arange_shape_like(y):
+    shape = y.shape
+    nelems = reduce(mul, shape)
+    x = nd.arange(nelems).reshape(shape)
+    return x
+
+
+class NDArrayGenerator(object):
+    def __init__(self, dim, startdim=1):
+        self.dim = dim
+        self.curdim = startdim
+
+    def __iter__(self):
+        return self
+
+    @staticmethod
+    def gen(dimensions):
+        shape = rand_shape_nd(dimensions, 4)
+        nelems = reduce(mul, shape)
+        x = nd.arange(nelems).reshape(shape)
+        return x
+
+    def next(self):
+        return self.__next__()
+
+    def __next__(self):
+        if self.curdim > self.dim:
+            raise StopIteration
+        x = NDArrayGenerator.gen(self.curdim)
+        self.curdim += 1
+        return x
+
+
+def flatten2d_right(x):
+    s_0 = x.shape[0]
+    s_1 = reduce(mul, x.shape[1:])
+    return x.reshape((s_0, s_1))
+
+
+def flatten2d_left(x):
+    s_0 = reduce(mul, x.shape[:-1])
+    s_1 = x.shape[-1]
+    return x.reshape((s_0, s_1))
+
+
+@with_seed()
+def test_dense_backward_flatten():
```

Comment:
Can we create a generic method to test gluon blocks such as `test_gluon_block(data, weight, block, expected)` just like we are doing with check_second_order_unary? It will be easier to verify if the test is correct.

Reply:
Why do we need this generic method? Seems to go against YAGNI. Where is the return on investment here?